fluid/utils.py

```diff
@@ -12,12 +12,59 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from compiler.ast import flatten
 
-def get_op_io_info(op):
-    inputs = dict([(name, op.input(name)) for name in op.input_names])
-    attrs = dict(
-        [(name, op.attr(name))
-         for name in op.attr_names]) if op.attr_names is not None else None
-    outputs = dict([(name, op.output(name)) for name in op.output_names])
 
-    return inputs, attrs, outputs
+class UniqOpIOs():
+    """Return input/output information for an operator, and resolve potential 
+       name conflicts in ONNX graph.
+    """
+
+    def __init__(self):
+        self._all_renamed_outputs = {}
+        self._renamed_cnt = 0
+
+    def get_new_name(self, arg):
+        """Get the new name for an argument.
+        """
+
+        self._renamed_cnt += 1
+        return arg + '@dup_' + str(self._renamed_cnt)
+
+    def rename_input_args(self):
+        """Rename input arguments if their previous output arugments have been 
+           renamed.
+        """
+
+        for in_name in self.inputs:
+            if self.inputs[in_name][0] in self._all_renamed_outputs:
+                self.inputs[in_name][0] = self._all_renamed_outputs[self.inputs[
+                    in_name][0]]
+
+    def rename_output_args(self):
+        """Rename output arguments if they have same names with the input 
+           arguments.
+        """
+
+        input_args = flatten(self.inputs.values())
+        for out_name in self.outputs:
+            if self.outputs[out_name][0] in input_args:
+                new_name = self.get_new_name(self.outputs[out_name][0])
+                self._all_renamed_outputs[self.outputs[out_name][0]] = new_name
+                self.outputs[out_name][0] = new_name
+
+    def __call__(self, op):
+        self.inputs = dict([(name, op.input(name)) for name in op.input_names])
+        self.attrs = dict(
+            [(name, op.attr(name))
+             for name in op.attr_names]) if op.attr_names is not None else None
+        self.outputs = dict(
+            [(name, op.output(name)) for name in op.output_names])
+
+        self.rename_input_args()
+        self.rename_output_args()
+
+        return self.inputs, self.attrs, self.outputs
+
+
+get_op_io_info = UniqOpIOs()
```

Comment:
So is this a object initialized for the entire program? Slightly uncommon pattern to do an initializer this way. Did you do it because you didn't want to change the `get_op_io_info` calls everywhere? I think it might be a good idea to initialize a class object to make these calls in place of this pattern

Reply:
Oh. Initialize `get_op_io_info` in this way mainly because we have a variable `self._all_renamed_outputs` here to store all the renamed outputs in the program, and the class `UniqOpIOs` can only be initialized once. Of course, we can make  the variable static:

```
class UniqOpIOs():
    _all_renamed_outputs = {}
    
    def __init__():
        ...
    def __call__():
        ...
```

And  `get_op_io_info` will be initialized like that:
```
get_op_io_info = UniqOpIOs()
inputs, attrs, outputs = get_op_io_info(op)
``` 
One more line code will be inserted whenever `get_op_io_info` is used. What do you   think?